File: Path.cpp
#include "Path.h"
#include <core/SkPoint.h>
#include "Util.h"

constexpr float EPS = 1e-3f;

void Path::addPoint(float x, float y, float r) {
  Point p{{x, y}, r};
  if (!points.empty()) {
    //check if previous point contains the new point
    //replace if the new point is larger
    if (points.back().contains(p)) {
      bool replace{};
      while (!points.empty() && points.back().r < r && points.back().contains(p)) {
        replace = true;
        points.pop_back();
      }
      if (!replace) return;
    }
  }
  points.push_back(p);
  dirty = true;
}

const SkPath &Path::getSkPath() {
  if (dirty) {
    computePath();
  }
  return path;
}

void Path::computePath() {
  dirty = false;
  path.rewind();

  const auto n = points.size();
  if (n == 1) {
    auto &p = points[0];
    path.addCircle(p.point.fX, p.point.fY, p.r);
  } else if (n > 1) {
    std::vector<LineSegment> segsF, segsB;
    for (size_t i = 0; i < n - 1; ++i) {
      auto &p = points[i];
      auto &q = points[i + 1];
      auto v = p.point - q.point;
      v.normalize();
      auto angle = tangentAngle(p, q);
      auto wF = rotate(v, angle);
      auto wB = rotate(v, -angle);
      segsF.push_back({p.point + wF * p.r, q.point + wF * q.r});
      segsB.push_back({q.point + wB * q.r, p.point + wB * p.r});
    }

    auto *seg0 = &segsB[0];
    path.moveTo(seg0->q);
    //printf("p.moveTo({%f, %f});\n", seg0->q.fX, seg0->q.fY);
    for (size_t i = 0; i < 2 * segsF.size(); ++i) {
      LineSegment *seg1;
      Point *p;
      if (i < segsF.size()) {
        seg1 = &segsF[i];
        p = &points[i];
      } else {
        auto j = i - segsF.size();
        seg1 = &segsB[segsB.size() - 1 - j];
        p = &points[n - 1 - j];
      }
      if (SkPoint intersect; i > 0 && computeIntersection(*seg0, *seg1, intersect)) {
        path.lineTo(intersect);
        //printf("p.lineTo({%f, %f});\n", intersect.fX, intersect.fY);
      } else {
        if (i > 0) path.lineTo(seg0->q);
        auto a0 = seg0->angle();
        auto a1 = seg1->angle() + 180;
        path.arcTo(p->bounds(), a0 - 90, 180 - (a0 - a1), false);
        //printf("p.arcTo({%f, %f, %f, %f}, %f, %f, %d);\n",
        //       p->bounds().fLeft, p->bounds().fTop, p->bounds().fRight, p->bounds().fBottom,
        //       a0 - 90, 180 - (a0 - a1), false);
      }
      seg0 = seg1;
    }
    path.close();
  }

}

float Path::tangentAngle(const Path::Point &p, const Path::Point &q) {
  auto d = (q.point - p.point).length();
  auto r = q.r - p.r; // d > r
  return std::acos(r / d);
}

bool Path::computeIntersection(const LineSegment &s1, const LineSegment &s2, SkPoint &i) {
  //https://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect
  const auto p0_x = s1.p.fX, p0_y = s1.p.fY;
  const auto p1_x = s1.q.fX, p1_y = s1.q.fY;
  const auto p2_x = s2.p.fX, p2_y = s2.p.fY;
  const auto p3_x = s2.q.fX, p3_y = s2.q.fY;

  const auto s1_x = p1_x - p0_x;
  const auto s1_y = p1_y - p0_y;
  const auto s2_x = p3_x - p2_x;
  const auto s2_y = p3_y - p2_y;

  const auto s = (-s1_y * (p0_x - p2_x) + s1_x * (p0_y - p2_y)) / (-s2_x * s1_y + s1_x * s2_y);
  const auto t = (s2_x * (p0_y - p2_y) - s2_y * (p0_x - p2_x)) / (-s2_x * s1_y + s1_x * s2_y);

  if (s >= 0 && s <= 1 && t >= 0 && t <= 1) {
    // Collision detected
    i = {p0_x + (t * s1_x), p0_y + (t * s1_y)};
    return true;
  }

  return false;
}

SkVector rotate(SkVector v, float angle) {
  auto cs = std::cos(angle);
  auto sn = std::sin(angle);
  return {
      v.fX * cs - v.fY * sn,
      v.fX * sn + v.fY * cs
  };
}

SkRect Path::Point::bounds() const {
  return SkRect::MakeXYWH(point.fX - r, point.fY - r, 2 * r, 2 * r);
}

bool Path::Point::contains(const Path::Point &o) const {
  return (point - o.point).length() < std::abs(r - o.r) + EPS;
}

float Path::LineSegment::angle() const {
  auto v = q - p;
  return rad2deg(std::atan2(v.fY, v.fX));
}

File: src/path.rs
use std::f64::consts::{FRAC_PI_2, PI, TAU};

use vello::kurbo::{Arc, BezPath, Circle, Line, Point, Shape, Vec2};

const EPS: f64 = 1e-3;

fn rotate(v: Vec2, angle: f64) -> Vec2 {
    let cs = angle.cos();
    let sn = angle.sin();
    Vec2 {
        x: v.x * cs - v.y * sn,
        y: v.x * sn + v.y * cs,
    }
}

#[derive(Default)]
pub struct PointPath {
    pub points: Vec<PointData>,
}

impl PointPath {
    pub fn new() -> Self {
        Self { points: Vec::new() }
    }

    pub fn add_point(&mut self, x: f64, y: f64, r: f64) {
        let p = PointData {
            point: Point { x, y },
            r,
        };
        while !self.points.is_empty() && p.contains(self.points.last().unwrap()) {
            self.points.pop();
        }
        if !self.points.is_empty() && self.points.last().unwrap().contains(&p) {
            return;
        }
        self.points.push(p);
    }

    // TODO does not look 100% right yet...
    pub fn bez_path(&mut self) -> BezPath {
        let n = self.points.len();
        if n == 0 {
            return BezPath::new();
        } else if n == 1 {
            let p = &self.points[0];
            let c = Circle::new(p.point, p.r);
            return c.to_path(0.1);
        }

        let mut segs_f = Vec::new();
        let mut segs_b = Vec::new();
        for i in 0..n - 1 {
            let p = &self.points[i];
            let q = &self.points[i + 1];
            let v = (p.point - q.point).normalize();
            let angle = tangent_angle(p, q);
            let w_f = rotate(v, angle);
            let w_b = rotate(v, -angle);
            segs_f.push(Line {
                p0: p.point + w_f * p.r,
                p1: q.point + w_f * q.r,
            });
            segs_b.push(Line {
                p0: q.point + w_b * q.r,
                p1: p.point + w_b * p.r,
            });
        }

        let mut seg0 = segs_b.first().unwrap();
        let mut path = BezPath::new();
        path.move_to(seg0.p1);
        for i in 0..2 * segs_f.len() {
            let (seg1, p) = if i < segs_f.len() {
                (&segs_f[i], &self.points[i])
            } else {
                let j = i - segs_f.len();
                (&segs_b[segs_b.len() - 1 - j], &self.points[n - 1 - j])
            };
            let mut is_intersection = false;
            if i > 0 {
                if let Some(intersect) = compute_intersection(seg0, seg1) {
                    path.line_to(intersect);
                    is_intersection = true;
                } else {
                    path.line_to(seg0.p1);
                }
            }
            if !is_intersection {
                let a0 = line_angle(seg0);
                let a1 = line_angle(seg1) + PI;
                let mut sweep = PI - (a0 - a1);
                if sweep >= PI {
                    sweep -= TAU;
                }
                let arc = Arc::new(p.point, (p.r, p.r), a0 - FRAC_PI_2, sweep, 0.);
                for b in arc.append_iter(0.1) {
                    path.push(b);
                }
            }
            seg0 = seg1;
        }
        path.close_path();
        path
    }

    pub fn reset(&mut self) {
        self.points.clear();
    }
}

#[derive(Debug)]
pub struct PointData {
    pub point: Point,
    pub r: f64,
}

impl PointData {
    fn contains(&self, other: &PointData) -> bool {
        let d = self.point.distance(other.point);
        d + other.r <= self.r + EPS
    }
}

fn tangent_angle(p: &PointData, q: &PointData) -> f64 {
    let d = p.point.distance(q.point);
    let r = q.r - p.r; // d > r
    (r / d).acos()
}

fn compute_intersection(s1: &Line, s2: &Line) -> Option<Point> {
    // https://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect
    let p0 = s1.p0;
    let p1 = s1.p1;
    let p2 = s2.p0;
    let p3 = s2.p1;

    let s1_x = p1.x - p0.x;
    let s1_y = p1.y - p0.y;
    let s2_x = p3.x - p2.x;
    let s2_y = p3.y - p2.y;

    let s = (-s1_y * (p0.x - p2.x) + s1_x * (p0.y - p2.y)) / (-s2_x * s1_y + s1_x * s2_y);
    let t = (s2_x * (p0.y - p2.y) - s2_y * (p0.x - p2.x)) / (-s2_x * s1_y + s1_x * s2_y);

    if s >= 0.0 && s <= 1.0 && t >= 0.0 && t <= 1.0 {
        let intersection: Point = Point {
            x: p0.x + (t * s1_x),
            y: p0.y + (t * s1_y),
        };
        Some(intersection)
    } else {
        None
    }
}

fn line_angle(l: &Line) -> f64 {
    let v = l.p1 - l.p0;
    v.y.atan2(v.x)
}
